### ISSUE TITLE
Bug with Unzip task.

### ISSUE THREAD

**Issue by dazinator** [Microsoft/extendvs#41]
Hi Guys.

I have found a bug in the msbuild task that ships in 

```
 <package id="Microsoft.VSSDK.BuildTools" version="15.0.26201" targetFramework="net461" developmentDependency="true" />
```

I first reported the issue here: https://social.msdn.microsoft.com/Forums/vstudio/en-US/7c4c9766-9ac4-4681-851d-7c69c4a73b57/problem-occurred-extracting-the-vsix-to-the-experimental-instance?forum=vsx

The problem, seems to be it's very fragile to file path lengths, but when it fails, it fails with a generic build output  error message that doesn't give any real detail as to what the problem is. I spent a whole day going through everything I could think of, including creating a console app to run the task directly agaisnt the vsix to repro the issue. In the end, when I shortened certain filenames (in my case the vstemplate file) the problem was solved.

**Comment by dazinator** [comment]
The reason why the vstemplate file is such an issue, is that when the VSIX is extracted, it's file name is used twice in the destination path.

For example, if you have a vstemplate file named: `MyProjectSystem.FooTemplate` then when this get's extracted the path is:


```
`C:\Users\Darrell\AppData\Local\Microsoft\VisualStudio\15.0_cd017506Exp\Extensions\MyProjectSystem.ProjectType\My Project System VSIX\1.8.3\ItemTemplates\My Project System\1033\MyProjectSystem.FooTemplate\MyProjectSystem.FooTemplate.vstemplate
```

So basically, if you are including an item template in your VSIX, and the template has a long name, you can very quickly exceed the 255 limit on windows (which I thought was going away) - which causes an issue with System.Packaging and the Unzip build task fails with no helpful error.

**Comment by dazinator** [comment]
IMHO there should be warnings for any file or manifest entry in the project that is both long in length and contributes to the length of the path on disc that gets extracted too. 

For example the `Product Name` and `Author` entered in the vsixmanifest file both appear to contribute to this path so should probably be kept as short as possible to mitigate this problem.

Also the     <ProjectType></ProjectType> entry in the vstemplate file for an Item template contributes to the path on disk, as well as the name of the manifest file itself.

**Comment by justcla** [comment]
Thanks for the feedback. It is most valuable. And your suggestions about the combined length of Product Name and Author are useful.
We'll look into this.

**Comment by yannduran** [comment]
I've just struck this same problem today, with a vsix project name that I've already used before (I'm basically recreating an existing project with exactly the same name).

This has not been a problem for me except for now.

@justcla Has the vsix installer been changed without us knowing or being notified it was being changed?

**Comment by dazinator** [comment]
As far as I could tell, the vsix installer itself is unaffected and works fine, and the problem is in the msbuild task - but this might have been fluke with the vsix installer as I installed my vsix via the installer fine, but then again the target directory would have been slightly shorter as when installing a vsix proper, it doesn't install it to the experimental instance directory (which is slightly longer).

**Comment by yannduran** [comment]
I'm not sure if this is related, but I'm also now having problems installing other people's extensions. I keep getting, basically, an error message that says an id has already be added, so it fails installation.

<rant>
It's getting really hard to work with VS 2017 as it stands right now, and again I have to protest that *VS Partners* are **not** getting access to early VS 2017 builds that are not released to the public (**under NDA of course**), like we got access to releases prior to the actual release of 2015 RTM. Instead, the RC versions are just being made available to **everyone** instead of just partners.

The idea was that partners would get to test their extension with as many releases as possible, both to check that their extensions work, and to report any bugs. 

How are we supposed to be able to **commit** to sim-shipping our extensions at the same time that VS 2017 ships, if we don't even know what the RTM bits (which I believe are now available internally) look like before the actual launch. It's **NUTS**. The way that the 2015 launch was handled was far better for partners than what's happened with 2017.
</rant>

I hate to think how many hours we partner *VS Partner Developers* are losing in productivity as a whole.

**Comment by dazinator** [comment]
That sounds like a seperate issue to me. My issue is very specific in nature. I understand exactly what you mean about time lost.. I've been dealing with asp.net core since beta, and the amount of tooling instability as the entire landscape has evolved has been a massive time sink. I am very much preying that we are over the worst now and tooling from MS will start to become more stable. He has had a few massive changes (like cps) so I don't expect the initial release of Vs2017 not to have any issues.

**Comment by yannduran** [comment]
Yeah, I agree it sounds like a separate issue, but it's still vsix-installer related, and started around the same time that I started experiencing the issue that you originally described. So I mentioned it, just in case.

**Comment by ivannaranjo** [comment]
This is also affecting me in a .vsix I work on when adding a reference to the `Microsoft.VisualStudio.Sdk.BuildTasks.14.0` package to be able to produce a VSIX V3 package from Visual Studio 2015. This means that it is part of the new tools for extensibility. Any idea if this is going to be look at soon?

**Comment by lzandman** [comment]
Also stuck on this issue. The VSIX is just fine, but deploying it to the experimental instance still fails. No fix in sight?

**Comment by justcla** [comment]
lzandman: Can you provide more information? What errors are you getting?
Does your VSIX contain very long file paths?

**Comment by lzandman** [comment]
I personally don't consider the paths very long, but I guess they're too long for Windows. Here's the relevant section (I think) of the output log:

```
2>Task "Unzip" (TaskId:156)
2>  Task Parameter:ZipFile=D:\git2\EDS.Shared\Templates\Rdw.Eds.FrontEnd.TemplateInstaller\Rdw.Eds.FrontEnd.TemplateInstaller\bin\Debug\Rdw.Eds.FrontEnd.TemplateInstaller.vsix (TaskId:156)
2>  Task Parameter:TargetDirectory=C:\Users\ZANDMANL\AppData\Local\Microsoft\VisualStudio\15.0_c3a435f5Exp\Extensions\EDS team\Rdw.Eds.FrontEnd.TemplateInstaller\1.0\ (TaskId:156)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\VSSDK\Microsoft.VsSDK.targets(693,5): error VSSDK1081: Problem occurred while extracting the vsix to the experimental extensions path. The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.
2>Done executing task "Unzip" -- FAILED. (TaskId:156)

```
I've manually extracted the VSIX package to that path and (of course) this also failed. It's hitting MAX_PATH issues. For now we've disabled extension deployment in our project settings (setting the 'DeployExtension' property to 'false'). But this makes testing the extension harder.

**Comment by bryanbcook** [comment]
@lzandman I found the same thing. Deselecting 'Deploy VSIX content to experimental instance for debugging' in the VSIX project will bypass the error and you'll have to manually deploy the extension to the experimental instance for testing.

Interestingly enough, I discovered a few really long file names when manually extracting the zip to the expected path. However, if I manually extracted the zip to a shorter path (C:\Temp) and then copied them into the folder, they work fine.

**Comment by goodtocode** [comment]
Please remember to apply the temporary workaround to all Configurations:
1. Right-click Project -> Properties
2. Click VSIX tab (on left)
3. Uncheck "Deploy VSIX content to experimental instance for debugging"
**Important:** Do this for all configurations

**Comment by fragmer** [comment]
The problem is still reproducible with VS 2019 16.3.3. Is this even a VSSDK problem, or are we just hitting known MSBuild limitations?

**Comment by dazinator** [comment]
@fragmer I've pinpointed the issue and the cause above. @justcla said they would look into it. Slightly disappointing to see its still an issue.
